### PR TITLE
Use binary search to find the min genesis block

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Improve the performance of calculating the genesis state from an eth1 chain when the minimum genesis time is somewhere in the middle of the chain by using binary search to find the first block that meets the time criteria.

### CODE REVIEW

File: pow/src/test/java/tech/pegasys/artemis/pow/MinimumGenesisTimeBlockFinderTest.java [hunk 1]
```diff
@@ -49,115 +46,69 @@ static void tearDown() {
...
+  public void shouldFindMinGenesisTimeBlockAtHeadOfChain() {
+    final Block[] blocks = withBlockTimestamps(4000, 5000, 6000, 7000, 8000);
+    final int minGenesisTime = 8000;
+    final Block expectedMinGenesisTimeBlock = blocks[blocks.length - 1];
+    assertMinGenesisBlock(blocks, minGenesisTime, expectedMinGenesisTimeBlock);
+  }
 
-    // since the genesis time calculated from the , we should publish the block
+  @Test
+  public void shouldFindMinGenesisTimeWithExactMatch() {
```
Comment (macfarla) [inline pow/src/test/java/tech/pegasys/artemis/pow/MinimumGenesisTimeBlockFinderTest.java:74]: this test is the same as shouldAddGenesisDelayToBlockTimestampWhenConsideringIfItIsTheMinGenesisBlock

File: pow/src/test/java/tech/pegasys/artemis/pow/MinimumGenesisTimeBlockFinderTest.java [hunk 2]
```diff
@@ -20,27 +20,24 @@
...
-  void setUp() {
-    eth1Provider = mock(Eth1Provider.class);
-
-    minimumGenesisTimeBlockFinder = new MinimumGenesisTimeBlockFinder(eth1Provider);
+  private MinimumGenesisTimeBlockFinder minimumGenesisTimeBlockFinder =
+      new MinimumGenesisTimeBlockFinder(eth1Provider);
 
+  @BeforeAll
+  static void setUp() {
+    // calculateCandidateGenesisTimestamp will return blockTime + 2
```
Comment (macfarla) [inline pow/src/test/java/tech/pegasys/artemis/pow/MinimumGenesisTimeBlockFinderTest.java]: I don't know if this comment is useful here?

**Review by macfarla — APPROVED** [review]
overall looks good. couple of minor things.

File: pow/src/test/java/tech/pegasys/artemis/pow/MinimumGenesisTimeBlockFinderTest.java (hunk 2 above)
Comment (ajsutton) [reply pow/src/test/java/tech/pegasys/artemis/pow/MinimumGenesisTimeBlockFinderTest.java]: It explains why the constant is being set and the effect it has on the genesis time derived from a block timestamp. I've tidied it up a bit.

File: pow/src/test/java/tech/pegasys/artemis/pow/MinimumGenesisTimeBlockFinderTest.java (hunk 1 above)
Comment (ajsutton) [reply pow/src/test/java/tech/pegasys/artemis/pow/MinimumGenesisTimeBlockFinderTest.java:74]: True, only need one of them.

File: pow/src/main/java/tech/pegasys/artemis/pow/MinimumGenesisTimeBlockFinder.java [hunk 3]
```diff
@@ -37,75 +42,40 @@ public MinimumGenesisTimeBlockFinder(Eth1Provider eth1Provider) {
...
+          .getEth1BlockFuture(mid)
+          .thenCompose(
+              midBlock -> {
+                final int cmp = compareBlockTimestampToMinGenesisTime(midBlock);
+                if (cmp < 0) {
+                  searchContext.low = mid.plus(ONE);
+                  return binarySearchLoop(searchContext);
+                } else if (cmp > 0) {
+                  if (mid.equals(ZERO)) {
+                    // The very first block is after min genesis so just use it
```
Comment (cemozerr) [inline pow/src/main/java/tech/pegasys/artemis/pow/MinimumGenesisTimeBlockFinder.java]:
```suggestion
                    // The very first block of eth1 is after eth 2 min genesis so just use it
```

**Review by cemozerr — APPROVED** [review]
Great improvement. Loved how you made the tests so legible.

File: pow/src/main/java/tech/pegasys/artemis/pow/MinimumGenesisTimeBlockFinder.java (hunk 3 above)
Comment (ajsutton) [reply pow/src/main/java/tech/pegasys/artemis/pow/MinimumGenesisTimeBlockFinder.java]: Good idea. Done.

**Pull request merged**